File: src/crew/hub.cpp
// hub.cpp -----------------------------------------------------------------------------------------------------

#include "crew/hub.h"

#include <chrono>
#include <cstring>
#include <iostream>

#if defined(_WIN32)
#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif
#include <winsock2.h>
#include <ws2tcpip.h>
#pragma comment(lib, "ws2_32.lib")
using SocketType = SOCKET;
constexpr SocketType InvalidSocket = INVALID_SOCKET;
#else
#include <arpa/inet.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <sys/socket.h>
#include <unistd.h>
using SocketType = int;
constexpr SocketType InvalidSocket = -1;
#define closesocket close
#endif

//-------------------------------------------------------------------------------------------------

namespace trellis::crew {

//-------------------------------------------------------------------------------------------------

static int RecvExact( SocketType s, char* buf, int len)
{
    int total = 0;
    while ( total < len) {
        int r = recv( s, buf + total, len - total, 0);
        if ( r <= 0) {
            return r;
        }
        total += r;
    }
    return total;
}

//-------------------------------------------------------------------------------------------------

static int SendExact( SocketType s, const char* buf, int len)
{
    int total = 0;
    while ( total < len) {
        int r = send( s, buf + total, len - total, 0);
        if ( r <= 0) {
            return r;
        }
        total += r;
    }
    return total;
}

//-------------------------------------------------------------------------------------------------

CrewHub::CrewHub()
{
}

//-------------------------------------------------------------------------------------------------

CrewHub::~CrewHub()
{
    Stop();
}

//-------------------------------------------------------------------------------------------------

void CrewHub::AddNode( uint32_t id, uint32_t mainPort, uint32_t asyncPort)
{
    std::lock_guard< std::mutex> lock( _HubMutex);
    _Nodes.push_back( std::make_shared< CrewNode>( id, mainPort, asyncPort));
}

//-------------------------------------------------------------------------------------------------

size_t CrewHub::NodeCount() const
{
    std::lock_guard< std::mutex> lock( _HubMutex);
    return _Nodes.size();
}

//-------------------------------------------------------------------------------------------------

std::shared_ptr< CrewNode> CrewHub::FindNode( uint32_t id) const
{
    std::lock_guard< std::mutex> lock( _HubMutex);
    for ( const auto& node : _Nodes) {
        if ( node->Id() == id) {
            return node;
        }
    }
    return nullptr;
}

//-------------------------------------------------------------------------------------------------

bool CrewHub::Start()
{
    if ( _IsRunning.load()) {
        return true;
    }

#if defined(_WIN32)
    WSADATA wsaData;
    if ( WSAStartup( MAKEWORD( 2, 2), &wsaData) != 0) {
        return false;
    }
#endif

    _IsRunning.store( true);

    std::lock_guard< std::mutex> lock( _HubMutex);
    for ( auto& node : _Nodes) {
        _Workers.emplace_back( &CrewHub::WorkerLoop, this, node);
    }

    return true;
}

//-------------------------------------------------------------------------------------------------

void CrewHub::Stop()
{
    if ( !_IsRunning.load()) {
        return;
    }
    _IsRunning.store( false);

    for ( auto& worker : _Workers) {
        if ( worker.joinable()) {
            worker.join();
        }
    }
    _Workers.clear();

#if defined(_WIN32)
    WSACleanup();
#endif
}

//-------------------------------------------------------------------------------------------------

bool CrewHub::IsNodeOnline( uint32_t id) const
{
    auto node = FindNode( id);
    if ( node != nullptr) {
        return node->IsOnline();
    }
    return false;
}

//-------------------------------------------------------------------------------------------------

NodeStats CrewHub::GetNodeStats( uint32_t id) const
{
    auto node = FindNode( id);
    if ( node != nullptr) {
        return node->GetStats();
    }
    return NodeStats{};
}

//-------------------------------------------------------------------------------------------------

void CrewHub::SetMessageCallback( MessageCallback cb)
{
    _MessageCb = cb;
}

//-------------------------------------------------------------------------------------------------

void CrewHub::WorkerLoop( std::shared_ptr< CrewNode> node)
{
    SocketType sMain = InvalidSocket;
    sockaddr_in addr;
    std::memset( &addr, 0, sizeof( addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons( static_cast< uint16_t>( node->MainPort()));
    inet_pton( AF_INET, "127.0.0.1", &addr.sin_addr);

    const int maxRetries = 100;
    for ( int retry = 0; retry < maxRetries && _IsRunning.load(); ++retry) {
        sMain = socket( AF_INET, SOCK_STREAM, IPPROTO_TCP);
        if ( sMain == InvalidSocket) {
            std::this_thread::sleep_for( std::chrono::milliseconds( 100));
            continue;
        }
        if ( connect( sMain, reinterpret_cast< sockaddr*>( &addr), sizeof( addr)) == 0) {
            break;
        }
        closesocket( sMain);
        sMain = InvalidSocket;
        std::this_thread::sleep_for( std::chrono::milliseconds( 100));
    }

    if ( sMain == InvalidSocket || !_IsRunning.load()) {
        return;
    }

    int noDelay = 1;
    setsockopt( sMain, IPPROTO_TCP, TCP_NODELAY, reinterpret_cast< const char*>( &noDelay), sizeof( noDelay));

    SocketType sAsync = InvalidSocket;
    sockaddr_in asyncAddr;
    std::memset( &asyncAddr, 0, sizeof( asyncAddr));
    asyncAddr.sin_family = AF_INET;
    asyncAddr.sin_port = htons( static_cast< uint16_t>( node->AsyncPort()));
    inet_pton( AF_INET, "127.0.0.1", &asyncAddr.sin_addr);

    for ( int retry = 0; retry < maxRetries && _IsRunning.load(); ++retry) {
        sAsync = socket( AF_INET, SOCK_STREAM, IPPROTO_TCP);
        if ( sAsync == InvalidSocket) {
            std::this_thread::sleep_for( std::chrono::milliseconds( 100));
            continue;
        }
        if ( connect( sAsync, reinterpret_cast< sockaddr*>( &asyncAddr), sizeof( asyncAddr)) == 0) {
            break;
        }
        closesocket( sAsync);
        sAsync = InvalidSocket;
        std::this_thread::sleep_for( std::chrono::milliseconds( 100));
    }

    if ( sAsync == InvalidSocket || !_IsRunning.load()) {
        closesocket( sMain);
        return;
    }

    // Renode Handshake
    ProtocolMessage hsReq;
    int r = RecvExact( sMain, reinterpret_cast< char*>( &hsReq), sizeof( hsReq));
    if ( r != sizeof( hsReq) || hsReq._ActionId != static_cast< int32_t>( CoSimAction::Handshake)) {
        closesocket( sMain);
        closesocket( sAsync);
        return;
    }

    ProtocolMessage hsResp;
    hsResp._ActionId = static_cast< int32_t>( CoSimAction::Handshake);
    hsResp._Addr = 0;
    hsResp._Value = 0;
    hsResp._PeripheralIndex = -1;
    SendExact( sMain, reinterpret_cast< const char*>( &hsResp), sizeof( hsResp));
    node->SetOnline( true);

    while ( _IsRunning.load()) {
        ProtocolMessage req;
        int rec = RecvExact( sMain, reinterpret_cast< char*>( &req), sizeof( req));
        if ( rec <= 0) {
            break;
        }

        if ( req._ActionId == static_cast< int32_t>( CoSimAction::Disconnect)) {
            break;
        }

        ProtocolMessage resp = HandleRequest( node, req);
        if ( SendExact( sMain, reinterpret_cast< const char*>( &resp), sizeof( resp)) <= 0) {
            break;
        }
    }

    node->SetOnline( false);
    closesocket( sMain);
    closesocket( sAsync);
}

//-------------------------------------------------------------------------------------------------

ProtocolMessage CrewHub::HandleRequest( std::shared_ptr< CrewNode> node, const ProtocolMessage& req)
{
    ProtocolMessage resp;
    resp._ActionId = static_cast< int32_t>( CoSimAction::Ok);
    resp._Addr = req._Addr;
    resp._Value = 0;
    resp._PeripheralIndex = req._PeripheralIndex;

    const uint64_t reg = req._Addr & 0xFFF;
    const auto action = static_cast< CoSimAction>( req._ActionId);

    if ( action == CoSimAction::ReadBus ||
         action == CoSimAction::ReadBusByte ||
         action == CoSimAction::ReadBusWord ||
         action == CoSimAction::ReadBusDword ||
         action == CoSimAction::ReadBusQword) {

        node->RecordRead();

        switch ( reg) {
        case REG_NODE_ID:
            resp._Value = static_cast< uint64_t>( node->Id());
            break;

        case REG_STATUS: {
            uint32_t status = STATUS_TX_READY;
            if ( node->RxCount() > 0) {
                status |= STATUS_RX_READY;
            }
            const uint32_t peerId = 1 - node->Id();
            if ( IsNodeOnline( peerId)) {
                status |= STATUS_PEER_UP;
            }
            resp._Value = static_cast< uint64_t>( status);
            break;
        }

        case REG_TX_DATA:
            resp._Value = 0;
            break;

        case REG_RX_DATA: {
            uint8_t byte = 0;
            if ( node->PopRx( byte)) {
                resp._Value = static_cast< uint64_t>( byte);
            } else {
                resp._Value = 0;
            }
            break;
        }

        case REG_RX_COUNT: {
            resp._Value = static_cast< uint64_t>( node->RxCount());
            break;
        }

        default:
            resp._Value = 0;
            break;
        }
    } else if ( action == CoSimAction::WriteBusByte ||
                action == CoSimAction::WriteBusWord ||
                action == CoSimAction::WriteBusDword ||
                action == CoSimAction::WriteBusQword) {

        node->RecordWrite();

        if ( reg == REG_TX_DATA) {
            const uint8_t byte = static_cast< uint8_t>( req._Value & 0xFF);
            node->RecordByteSent();

            const uint32_t peerId = 1 - node->Id();
            auto peer = FindNode( peerId);
            if ( peer != nullptr) {
                peer->PushRx( byte);
                if ( _MessageCb) {
                    _MessageCb( node->Id(), peerId, byte);
                }
            }
        }
    } else if ( action == CoSimAction::ResetPeripheral) {
        node->ClearRx();
    }

    return resp;
}

} // namespace trellis::crew

File: src/crew/hub.rs
//-- hub.rs ---------------------------------------------------------------------------------------
//--------------------------------------------------------------------------------------------------

use	crate::crew::config::CrewLinkConfig;
use	crate::crew::node::{ CrewNode, NodeStats };
use	crate::crew::protocol::{ CoSimAction, ProtocolMessage, REG_NODE_ID, REG_RX_COUNT, REG_RX_DATA, REG_STATUS, REG_TX_DATA, STATUS_PEER_UP, STATUS_RX_READY, STATUS_TX_READY };
use	crate::silo::buff::Buff;
use	crate::silo::stash::Stash;
use	crate::silo::useg::USeg;
use	crate::stalks::work::SpinMutex;
use	std::sync::Arc;

//--------------------------------------------------------------------------------------------------
/// Callback type for monitoring byte-level routing between VM nodes.
pub type MessageCallback = Arc< dyn Fn( u32, u32, u8) + Send + Sync>;

//--------------------------------------------------------------------------------------------------
/// Concrete co-simulation hub managing VM nodes and in-memory protocol dispatch.
pub struct CrewHub
{
    _Nodes: SpinMutex< Stash< Arc< CrewNode>>>,
    _MessageCb: SpinMutex< Option< MessageCallback>>,
    _Routes: SpinMutex< Stash< ( u32, Buff< u32>)>>,
}
impl Default for CrewHub {
    fn	default() -> Self
    {
        Self::New()
    }
}
impl CrewHub
{
    pub fn	New() -> Self
    {
        Self {
            _Nodes: SpinMutex::New( Stash::New()),
            _MessageCb: SpinMutex::New( None),
            _Routes: SpinMutex::New( Stash::New()),
        }
    }
    #[inline]
    pub fn	new() -> Self
    {
        Self::New()
    }
    pub fn	AddLink( &self, config: CrewLinkConfig)
    {
        let  	mut routes = self._Routes.Lock();
        let  	mut found = false;
        let  	src = config.source_node_id;
        USeg::FromLen( routes.Size()).Traverse( |i| {
            if routes[i].0 == src {
                found = true;
            }
        });
        if !found {
            routes.Push( ( src, config.destination_node_ids));
        }
    }
    #[inline]
    pub fn	add_link( &self, config: CrewLinkConfig)
    {
        self.AddLink( config);
    }
    pub fn	AddNode( &self, id: u32)
    {
        let  	mut nodes = self._Nodes.Lock();
        nodes.Push( Arc::new( CrewNode::New( id)));
    }
    #[inline]
    pub fn	add_node( &self, id: u32)
    {
        self.AddNode( id);
    }
    pub fn	NodeCount( &self) -> u32
    {
        let  	nodes = self._Nodes.Lock();
        nodes.Size()
    }
    #[inline]
    pub fn	node_count( &self) -> u32
    {
        self.NodeCount()
    }
    pub fn	FindNode( &self, id: u32) -> Option< Arc< CrewNode>>
    {
        let  	nodes = self._Nodes.Lock();
        let  	mut found: Option< Arc< CrewNode>> = None;
        USeg::FromLen( nodes.Size()).Traverse( |i| {
            if found.is_none() && nodes[i].Id() == id {
                found = Some( nodes[i].clone());
            }
        });
        found
    }
    #[inline]
    pub fn	find_node( &self, id: u32) -> Option< Arc< CrewNode>>
    {
        self.FindNode( id)
    }
    pub fn	IsNodeOnline( &self, id: u32) -> bool
    {
        if let  	Some( node) = self.FindNode( id) {
            node.IsOnline()
        } else {
            false
        }
    }
    #[inline]
    pub fn	is_node_online( &self, id: u32) -> bool
    {
        self.IsNodeOnline( id)
    }
    pub fn	GetNodeStats( &self, id: u32) -> NodeStats
    {
        if let  	Some( node) = self.FindNode( id) {
            node.GetStats()
        } else {
            NodeStats::default()
        }
    }
    #[inline]
    pub fn	get_node_stats( &self, id: u32) -> NodeStats
    {
        self.GetNodeStats( id)
    }
    pub fn	SetMessageCallback< F>( &self, cb: F)
    where
        F: Fn( u32, u32, u8) + Send + Sync + 'static,
    {
        let  	mut cbGuard = self._MessageCb.Lock();
        *cbGuard = Some( Arc::new( cb));
    }
    #[inline]
    pub fn	set_message_callback< F>( &self, cb: F)
    where
        F: Fn( u32, u32, u8) + Send + Sync + 'static,
    {
        self.SetMessageCallback( cb);
    }
    pub fn	HandleRequest( &self, node: &CrewNode, req: &ProtocolMessage) -> ProtocolMessage
    {
        let  	mut resp = ProtocolMessage {
            _ActionId: CoSimAction::Ok as i32,
            _Addr: req.Addr(),
            _Value: 0,
            _PeripheralIndex: req.PeripheralIndex(),
        };
        let  	reg = ( req.Addr() & 0xFFF) as u32;
        let  	action = req.Action();
        match action {
            CoSimAction::ReadBus
            | CoSimAction::ReadBusByte
            | CoSimAction::ReadBusWord
            | CoSimAction::ReadBusDword
            | CoSimAction::ReadBusQword => {
                node.RecordRead();
                match reg {
                    REG_NODE_ID => {
                        resp._Value = node.Id() as u64;
                    }
                    REG_STATUS => {
                        let  	mut status = 0;
                        if node.RxCount() > 0 {
                            status |= STATUS_RX_READY;
                        }
                        let  	peers = self.getPeersForNode( node.Id());
                        let  	mut allReady = !peers.IsEmpty();
                        let  	mut anyOnline = false;
                        USeg::FromLen( peers.Len()).Traverse( |i| {
                            if let  	Some( peer) = self.FindNode( peers[i]) {
                                if peer.IsOnline() {
                                    anyOnline = true;
                                } else {
                                    allReady = false;
                                }
                                if !peer.CanPushRx() {
                                    allReady = false;
                                }
                            } else {
                                allReady = false;
                            }
                        });
                        if allReady {
                            status |= STATUS_TX_READY;
                        }
                        if anyOnline {
                            status |= STATUS_PEER_UP;
                        }
                        resp._Value = status as u64;
                    }
                    REG_TX_DATA => {
                        resp._Value = 0;
                    }
                    REG_RX_DATA => {
                        let  	mut byte = 0u8;
                        if node.PopRx( &mut byte) {
                            resp._Value = byte as u64;
                        } else {
                            resp._Value = 0;
                        }
                    }
                    REG_RX_COUNT => {
                        resp._Value = node.RxCount() as u64;
                    }
                    _ => {
                        resp._Value = 0;
                    }
                }
            }
            CoSimAction::WriteBusByte
            | CoSimAction::WriteBusWord
            | CoSimAction::WriteBusDword
            | CoSimAction::WriteBusQword => {
                node.RecordWrite();
                if reg == REG_TX_DATA && action == CoSimAction::WriteBusByte {
                    let  	byte = ( req.Value() & 0xFF) as u8;
                    let  	peers = self.getPeersForNode( node.Id());
                    let  	mut accepted = !peers.IsEmpty();
                    USeg::FromLen( peers.Len()).Traverse( |i| {
                        let  	peerId = peers[i];
                        if let  	Some( peer) = self.FindNode( peerId) {
                            if !peer.IsOnline() || !peer.CanPushRx() {
                                accepted = false;
                            }
                        } else {
                            accepted = false;
                        }
                    });
                    if !accepted {
                        resp._ActionId = CoSimAction::Error as i32;
                        return resp;
                    }
                    node.RecordByteSent();
                    let  	mut allDelivered = true;
                    USeg::FromLen( peers.Len()).Traverse( |i| {
                        let  	peerId = peers[i];
                        if let  	Some( peer) = self.FindNode( peerId) {
                            if peer.TryPushRx( byte) != crate::crew::node::RxPushOutcome::Delivered {
                                allDelivered = false;
                            } else {
                                let  	cbOpt = {
                                    let  	guard = self._MessageCb.Lock();
                                    guard.clone()
                                };
                                if let  	Some( cb) = cbOpt {
                                    cb( node.Id(), peerId, byte);
                                }
                            }
                        }
                    });
                    if !allDelivered {
                        resp._ActionId = CoSimAction::Error as i32;
                    }
                } else {
                    resp._ActionId = CoSimAction::Error as i32;
                }
            }
            CoSimAction::ResetPeripheral => {
                node.ClearRx();
            }
            _ => {}
        }
        resp
    }
    #[inline]
    pub fn	handle_request( &self, node: &CrewNode, req: &ProtocolMessage) -> ProtocolMessage
    {
        self.HandleRequest( node, req)
    }
    fn	getPeersForNode( &self, nodeId: u32) -> Buff< u32>
    {
        let  	routes = self._Routes.Lock();
        let  	mut dests = Buff::New();
        USeg::FromLen( routes.Size()).Traverse( |i| {
            if routes[i].0 == nodeId {
                dests = routes[i].1.clone();
            }
        });
        if dests.IsEmpty() {
            // Trellis default: peerId = 1 - node.Id()
            let  	peerId = if nodeId == 0 { 1 } else { 0 };
            if self.FindNode( peerId).is_some() {
                dests = Buff::FromDispenser( 1, |_| peerId);
            }
        }
        dests
    }
}
